File: backend/app/schemas/simulation.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class SimulationCreate(BaseModel):
    protocol: str = Field(
        min_length=1,
        max_length=50,
        examples=["tcp"],
    )

    input_parameters: dict[str, Any] = Field(
        default_factory=dict,
    )


class SimulationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    scenario_id: int | None
    protocol: str
    input_parameters: dict[str, Any]
    result: dict[str, Any] | None
    status: str
    created_at: datetime


class SimulationListResponse(BaseModel):
    items: list[SimulationResponse]
    total: int
    skip: int
    limit: int

File: frontend/types/simulation.ts
export type TcpEndpoint = "client" | "server";

export type TcpHandshakeEvent = {
  step: number;
  sent_at_ms: number;
  arrives_at_ms: number;
  source: TcpEndpoint;
  destination: TcpEndpoint;
  flags: string[];
  sequence_number: number;
  acknowledgment_number: number | null;
  client_state: string;
  server_state: string;
  description: string;
};

export type TcpHandshakeResult = {
  handshake: string;
  events: TcpHandshakeEvent[];
  final_state: {
    client: string;
    server: string;
  };
  connection_established: boolean;
  total_duration_ms: number;
};

export type SimulationRun = {
  id: number;
  scenario_id: number | null;
  protocol: string;
  input_parameters: Record<string, number>;
  result: TcpHandshakeResult | null;
  status: string;
  created_at: string;
};

export type SimulationCreateRequest = {
  protocol: string;
  input_parameters: Record<string, number>;
};

export type SimulationListResponse = {
  items: SimulationRun[];
  total: number;
  skip: number;
  limit: number;
};